### ISSUE TITLE
`RuntimeError` using `torchmetrics.IoU`

### ISSUE THREAD

**Issue by Nora-Go** [Lightning-AI/torchmetrics#232]
Hi :)

I get this error using torchmetrics.IoU(num_classes=4) and running my model on gpu: 
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!

If I write my own IoU (using some code of yours and simplifying it), everything works perfectly well.

It would be great if you could resolve this issue :) 
Thanks for your time and help!

**Comment by SkafteNicki** [comment]
I am going to take a guess here, but as @carmocca mentions a reproducible script/snippet would be nice :]

We commonly see this error when the metric is not correctly recognized inside the LightningModule. This section describes in details how metrics should be initialized inside the lightningmodule:
https://torchmetrics.readthedocs.io/en/latest/pages/overview.html#metrics-and-devices
@Nora-Go could that be your problem?

**Comment by Nora-Go** [comment]
Hi :) sorry this took so long. You can find the hopefully minimal example in the attached file. 
As suggested in https://torchmetrics.readthedocs.io/en/latest/pages/overview.html#metrics-and-devices, I checked wether the metric is in the children() of the module and it is! So I guess that's not the problem.

Please let me know if the minimal example is not good (it's my first try on a minimal example ;) ), and how I should change it to make it better.


[MinimalExample.zip](https://github.com/PyTorchLightning/metrics/files/6449982/MinimalExample.zip)

**Comment by SkafteNicki** [comment]
@Nora-Go thanks for response. I try to reproduce and debug with the example you send :]

**Comment by Nora-Go** [comment]
Ok I messed up the print out of children()... (Sorry about that).
I fixed it by using:

`self.metric = torchmetrics.IoU(num_classes=4)`
`hparams['metric'] = self.metric`

(Yes I know that looks weird, but I need the metric to be inside of hparams, so that it gets logged...
Using self.save_hyperparameters() just is confusing, but mandatory in the newest release of pytorch lightning.)

However, now it works! Thanks for your patience with me!
